### PR TITLE
Some random changes

### PR DESCRIPTION
**Pull request by Malacath-92**
- add return value to `spaw_tree`
- add `spawn_unrolled_player_rope`
- add `set_entity_climbable`
- add `set_entity_turning_enabled`
- split `entity.hpp/.cpp` into `entity_db.cpp/.hpp`, `entity_hooks_info.hpp`, `entity_hooks_structs.hpp` and `entity.hpp/.cpp`
- consolidate all collision info into one struct (guided by usage in the binary)
- fix a bug with vanilla machine rooms
- fix wrong param passed in `update_movable`
- expose `CustomMovableBehavior::base_behavior`
- add a string-literal operator `""_gh` for Ghdira byte strings
- add some utilities for `PatternCommandBuffer`
- add interface to disable all hooks

### CODE REVIEW

**Review by Dregu — APPROVED** [review]

File: src/game_api/entity_hooks_info.hpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    std::vector<HookWithId<bool(Entity*)>> on_player_instagib;
+    std::vector<HookWithId<bool(Entity*, Entity*, int8_t, float, float, uint16_t, uint8_t)>> on_damage;
+    std::vector<HookWithId<bool(Movable*)>> pre_statemachine;
+    std::vector<HookWithId<void(Movable*)>> post_statemachine;
+    std::vector<HookWithId<void(Container*, Movable*)>> on_open;
+    std::vector<HookWithId<bool(Entity*, Entity*)>> pre_collision1;
+    std::vector<HookWithId<bool(Entity*, Entity*)>> pre_collision2;
+    std::vector<HookWithId<bool(Entity*)>> pre_render;
+    std::vector<HookWithId<void(Entity*)>> post_render;
+};
```
Comment (Mr-Auto) [inline src/game_api/entity_hooks_info.hpp:36]: no empty line

**Comment by Dregu** [comment]
Climb code would probably easily bend to any properties flag per entity. It's already crazy!

File: src/game_api/movable_behavior.cpp [hunk 2]
```diff
@@ -190,36 +194,213 @@ void update_movable(Movable* movable, bool disable_gravity)
...
     DetourTransactionBegin();
     DetourUpdateThread(GetCurrentThread());
 
     auto memory = Memory::get();
-    g_update_movable_trampoline = (UpdateMovable*)memory.at_exe(0x228e3580);
 
+    g_get_entity_close_to_trampoline = (GetEntityCloseTo*)get_address("layer_get_entity_close_to"sv);
+    DetourAttach((void**)&g_get_entity_close_to_trampoline, (GetEntityCloseTo*)get_entity_close_to);
+
+    g_entity_turn_trampoline = (EntityTurn*)get_address("entity_turn"sv);
```
Comment (Mr-Auto) [inline src/game_api/movable_behavior.cpp:388]: could be replaced with `get_virtual_function_address` to not add another pattern to search for at the start

File: src/game_api/entity.hpp [hunk 3]
```diff
@@ -1,444 +1,249 @@
...
+
+std::tuple<float, float, uint8_t> get_position(uint32_t uid);
+std::tuple<float, float, uint8_t> get_render_position(uint32_t uid);
+
+std::tuple<float, float> get_velocity(uint32_t uid);
+
+AABB get_hitbox(uint32_t uid, bool use_render_pos);
+
+Entity* get_entity_ptr(uint32_t uid);
+Entity* get_entity_ptr_local(uint32_t uid);
```
Comment (Mr-Auto) [inline src/game_api/entity.hpp:249]: github doesn't know how to chose the changes here, was this just moving out the entity db and other structs to a different file?

File: src/game_api/entity.cpp [hunk 4]
```diff
@@ -1,898 +1,812 @@
...
+        {
+            original(ent, gravity);
+        },
+        0x53);
+}
+
+void Movable::reset_gravity()
+{
+    unregister_hook_function((void***)this, 0x53);
+}
```
Comment (Mr-Auto) [inline src/game_api/entity.cpp:812]: What's changed in this file? github doesn't show it well

**Review by Mr-Auto — CHANGES_REQUESTED** [review]

File: src/game_api/entity_db.hpp [hunk 5]
```diff
@@ -99,6 +99,9 @@ struct EntityDB
     float default_special_offsetx;
     float default_special_offsety;
     uint8_t init;
+
+    EntityDB(const EntityDB& other) = default;
```
Comment (Dregu) [inline src/game_api/entity_db.hpp:103]: This won't copy animations and in fact probably crashes with animations, but that's for someone else to finish. Works great for wallclimbing though:

```lua
E.type = EntityDB:new(E.type) -- create a new type with the copy constructor
E.type.properties_flags = set_flag(E.type.properties_flags, 6) -- set only this entity climbable
```

Comment (Malacath-92) [reply src/game_api/entity_db.hpp:103]: Why will this not copy animations?

File: src/game_api/entity.cpp (hunk 4 above)
Comment (Malacath-92) [reply src/game_api/entity.cpp:812]: Only the facing left and climbable functions.

Comment (Malacath-92) [reply src/game_api/entity.cpp:812]: And moving out stuff to the other files (mentioned in the description)

File: src/game_api/entity.hpp (hunk 3 above)
Comment (Malacath-92) [reply src/game_api/entity.hpp:249]: Same as the .cpp

File: src/game_api/movable_behavior.cpp (hunk 2 above)
Comment (Malacath-92) [reply src/game_api/movable_behavior.cpp:388]: Right, you mentioned before, I forgot about that

**Review by Dregu — APPROVED** [review]

**Pull request merged**